### PR TITLE
[Fix] Enabling Megatron FSDP run with TE FP8

### PR DESCRIPTION
**Pull request by AllenFarcas**
## Motivation

Enabling Megatron-LM custom FSDP training with two modes: FP8 primary weights, replacing BF16/FP16 weights and another one for fp8_autocast with BF16 weights and FP8 compute cache

## Technical Details

Fixes issue https://github.com/ROCm/frameworks-internal/issues/13927

## Test Plan

Ran tests in TE, Ran llama3 examples with the new changes.

## Submission Checklist

- [x] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.

### CODE REVIEW

File: examples/llama/train_llama3.sh [hunk 1]
```diff
@@ -85,11 +85,21 @@ SAVE_INTERVAL="${SAVE_INTERVAL:-5000}"
 EVAL_ITERS="${EVAL_ITERS:-'-1'}"
 CKPT_FORMAT="${CKPT_FORMAT:-torch}"
 DATA_CACHE_PATH="${DATA_CACHE_PATH:-/root/cache}"
+MEGATRON_FSDP="${MEGATRON_FSDP:-0}"
```
Comment (wenchenvincent) [inline examples/llama/train_llama3.sh:88]: Also apply the changes to train_llama2.sh

File: examples/llama/README.md [hunk 2]
```diff
@@ -16,33 +16,24 @@ This guide provides the steps for setting up the environment and configuring the
...
-### 2.2 Multi-node Training
-To run training on multiple nodes, launch the Docker container on each node. Follow these steps:
+3. **Install Megatron-LM**
+  Run  `pip install .` in `/workspace/Megatron-LM` to install megatron package.
 
-- **On the Master Node:**
-  ```bash
-  TEE_OUTPUT=1 MBS=2 BS=64 TP=8 TE_FP8=0 SEQ_LENGTH=4096 bash examples/llama/train_llama2.sh
-  ```
+    **Note** that it is also possible to use `rocm/megatron-lm:latest` like images, which
```
Comment (wenchenvincent) [inline examples/llama/README.md]: I am not sure if `rocm/megatron-lm:latest` has Megatron installed. It surely has the working copy of Megatron repo in it. Could you confirm?

File: examples/llama/README.md [hunk 3]
```diff
@@ -16,33 +16,24 @@ This guide provides the steps for setting up the environment and configuring the
...
-
-## 2. How to Run
-
-### 2.1 Single Node Training
-To run the training on a single node, go to Megatron-LM folder, use the following command:
-```bash
-TEE_OUTPUT=1 MBS=2 BS=64 TP=8 TE_FP8=0 SEQ_LENGTH=4096 bash examples/llama/train_llama2.sh
-```
-
+    **Note** that it is recommended to use `rocm/pytorch-training:latest` like images which
```
Comment (wenchenvincent) [inline examples/llama/README.md]: it is recommended to use docker images like `rocm/pytorch-training:latest`, which
have most requirements setup (e.g., `PyTorch >= 2.5.0` is needed for full support
of FSDP-v2).

**Comment by wenchenvincent** [comment]
Also, please change the title of the PR to be more descriptive.

File: examples/llama/README.md [hunk 4]
```diff
@@ -63,15 +54,49 @@ You can use either mock data or real data for training.
...
+When preparing a dataset, a tokenizer is required. The scripts support tokenizers
+which are fully specified with choices of `TOKENIZER_TYPE` and `TOKENIZER_MODEL`. With
+the exception of Llama 2 training script, the default `TOKENIZER_TYPE` is
+`HuggingFaceTokenizer` and for it, only a valid `TOKENIZER_MODEL` is needed. For
+example, after obtaining a [permission](https://huggingface.co/meta-llama/Llama-3.1-8B),
+run
 
-### 3.3 Tokenizer
+```bash
+wget --header="Authorization: Bearer $HF_TOKEN" -O tokenizer/special_tokens_map.json https://huggingface.co/meta-llama/Llama-3.1-8B/resolve/main/special_tokens_map.json
```
Comment (sudhu2k) [inline examples/llama/README.md]: I think when --tokenizer-type=HuggingFaceTokenizer, huggingface will automatically download the tokenizer files when specifying --tokenizer-model with the model id (e.g. meta-llama/Llama-3-8B). So no need for manual download. But the user would have to set HF_TOKEN if it's a restricted tokenzier.

File: examples/llama/train_llama3.sh [hunk 5]
```diff
@@ -322,6 +332,16 @@ if [ "$TE_FP8" -eq 1 ]; then
         exit
     fi
 
+    if [ "$FSDP" -eq 1 ]; then
+        EXTRA_ARGS="$EXTRA_ARGS --keep_fp8_weight_transpose_cache"
```
Comment (sudhu2k) [inline examples/llama/train_llama3.sh]: When using pytorch FSDP2, we don't necessarily have to force to keep fp8 transpose cache. Because with the latest changes, for delayed/current scaling we only perform all gather on the rowwise data, hence TE force sets fp8 weight transpose cache to false anyways when pytorch FSDP is enabled. But this is not the case for MXFP8, for MXFP8 both rowwise data/scale and columnwise data/scales are being all gathered, so TE force sets the keep fp8 transpose cache to True when using pytorch FSDP2. So we can safely remove this.

**Review by sudhu2k — CHANGES_REQUESTED** [review]

File: examples/llama/README.md (hunk 2 above)
Comment (AllenFarcas) [reply examples/llama/README.md]: Did double check and `rocm/megatron-lm:latest` doesn't have Megatron installed. It does have the `Megatron-LM` under `/workspace` but the user must run `pip install .` to actually have it installed.

File: examples/llama/train_llama3.sh [hunk 6]
```diff
@@ -322,6 +332,12 @@ if [ "$TE_FP8" -eq 1 ]; then
         exit
     fi
 
+    if [ "$MEGATRON_FSDP" -eq 1 ]; then
+        EXTRA_ARGS="$EXTRA_ARGS --keep_fp8_weight_transpose_cache" 
+        if [ "$FP8_PARAM_GATHER" -eq 1 ]; then
```
Comment (sudhu2k) [inline examples/llama/train_llama3.sh]: The fp8 param gather isn't exclusive for megatron fsdp, so probably we can keep it outside the if case.

File: examples/llama/train_llama2.sh [hunk 7]
```diff
@@ -339,8 +346,11 @@ if [ "$TE_FP8" -eq 1 ]; then
         exit
     fi
 
-    if [ "$FSDP" -eq 1 ]; then
+    if [ "$MEGATRON_FSDP" -eq 1 ]; then
         EXTRA_ARGS="$EXTRA_ARGS --keep_fp8_weight_transpose_cache" 
+        if [ "$FP8_PARAM_GATHER" -eq 1 ]; then
```
Comment (sudhu2k) [inline examples/llama/train_llama2.sh]: Pasting same comment  here:
The fp8 param gather isn't exclusive for megatron fsdp, so probably we can keep it outside the if case.

**Review by sudhu2k — CHANGES_REQUESTED** [review]

**Review by sudhu2k — APPROVED** [review]
LGTM.

File: examples/llama/train_llama3.sh [hunk 8]
```diff
@@ -349,6 +343,16 @@ else
    LOGGING_ARGS=""
 fi
 
+if [ "$MEGATRON_FSDP" -eq 1 ]; then
+    EXTRA_ARGS="$EXTRA_ARGS --use-megatron-fsdp --ckpt-format fsdp_dtensor --data-parallel-sharding-strategy optim_grads_params --use-distributed-optimizer --use-nccl-ub --overlap-grad-reduce --overlap-param-gather --sequence-parallel"
```
Comment (wenchenvincent) [inline examples/llama/train_llama3.sh]: Why use sequence parallel with fsdp?

File: examples/llama/train_llama3.sh [hunk 9]
```diff
@@ -333,6 +342,15 @@ else
    LOGGING_ARGS=""
 fi
 
+if [ "$MEGATRON_FSDP" -eq 1 ]; then
```
Comment (wenchenvincent) [inline examples/llama/train_llama3.sh:345]: I think you will need `unset CUDA_DEVICE_MAX_CONNECTIONS` for megatron fsdp as well.

File: examples/llama/train_llama3.sh (hunk 8 above)
Comment (AllenFarcas) [reply examples/llama/train_llama3.sh]: Removed, was from older attempts to see if some flags add to the performance.

File: examples/llama/train_llama3.sh (hunk 9 above)
Comment (AllenFarcas) [reply examples/llama/train_llama3.sh:345]: Fixed.

File: examples/llama/README.md [hunk 10]
```diff
@@ -219,7 +219,7 @@ follow these steps for 2 Node run with Node0 as master node :
   `1` to enable Megatron-LM's custom FSDP with DTensor checkpointing (default: 0). It adds automatically `--use-megatron-fsdp --ckpt-format fsdp_dtensor` in the script. Of note, this disables `TP>1` automatically.
 
 - **CUDA_DEVICE_MAX_CONNECTIONS**
```
Comment (wenchenvincent) [inline examples/llama/README.md]: I think we can just remove the mentioning of CUDA_DEVICE_MAX_CONNECTIONS in the README.

File: examples/llama/README.md [hunk 11]
```diff
@@ -102,34 +116,125 @@ If you're running multi-node training, update the following environment variable
...
+use the following command:
+```bash
+TEE_OUTPUT=1 MBS=2 BS=16 TP=1 TE_FP8=0 FSDP=1 RECOMPUTE=1 SEQ_LENGTH=8192 MODEL_SIZE=8 bash examples/llama/train_llama3.sh
+```
+**Note:** It is suggested to use `TP=1` when FSDP is enabled, for higher throughput.
+And FSDP-v2 is not supported with pipeline parallelism, expert parallelism, MCore's
+distributed optimizer, gradient accumulation fusion and fp16.
+
+#### FP8 options with Megatron-LM FSDP (train_llama3.sh)
+  - **FP8 primaries (fp8_model_init, param gather):** `TE_FP8=1 MEGATRON_FSDP=1 FP8_PARAM_GATHER=1 bash examples/llama/train_llama3.sh`
```
Comment (wenchenvincent) [inline examples/llama/README.md]: primaries -> primary weights

File: examples/llama/train_llama2.sh [hunk 12]
```diff
@@ -339,12 +344,20 @@ if [ "$TE_FP8" -eq 1 ]; then
         exit
     fi
 
-    if [ "$FSDP" -eq 1 ]; then
+    if [ "$FP8_PARAM_GATHER" -eq 1 ]; then
+        EXTRA_ARGS="$EXTRA_ARGS --fp8-param-gather" 
+    fi
+
+    if [ "$MEGATRON_FSDP" -eq 1 ]; then
         EXTRA_ARGS="$EXTRA_ARGS --keep_fp8_weight_transpose_cache"
```
Comment (wenchenvincent) [inline examples/llama/train_llama2.sh]: It seems that this need to be corrected to `--keep-fp8-weight-transpose-cache`

File: examples/llama/train_llama3.sh [hunk 13]
```diff
@@ -322,6 +324,13 @@ if [ "$TE_FP8" -eq 1 ]; then
         exit
     fi
 
+    if [ "$FP8_PARAM_GATHER" -eq 1 ]; then
+        EXTRA_ARGS="$EXTRA_ARGS --fp8-param-gather" 
+    fi
+
+    if [ "$MEGATRON_FSDP" -eq 1 ]; then
+        EXTRA_ARGS="$EXTRA_ARGS --keep_fp8_weight_transpose_cache"
```
Comment (wenchenvincent) [inline examples/llama/train_llama3.sh]: It seems that this need to be corrected to `--keep-fp8-weight-transpose-cache`

Comment (wenchenvincent) [reply examples/llama/train_llama3.sh]: @AllenFarcas It should be `--keep-fp8-weight-transpose-cache` instead of `--keep-fp8-transpose-cache`. Please fix other places as well.

Also, please don't mark the comments as resolved by yourself. It is the reviewer's responsibility to mark the comments as resolved.

Comment (AllenFarcas) [reply examples/llama/train_llama3.sh]: @wenchenvincent  In `megatron/training/arguments.py` there are only two arguments related to transpose cache: `--keep_fp8_weight_transpose_cache` and `--keep-fp8-transpose-cache`. I believe we want the version with the weight transpose cache. Should I change the argument in the `arguments.py` to be `--keep-fp8-weight-transpose-cache` instead of how it is now, i.e., ` --keep_fp8_weight_transpose_cache`?

Comment (wenchenvincent) [reply examples/llama/train_llama3.sh]: `--keep_fp8_weight_transpose_cache` is what we implemented before for ROCm. `--keep-fp8-transpose-cache` is from upstream. We need to investigate whether the two are for the same purpose and whether they can be consolidated.

**Review by wenchenvincent — APPROVED** [review]
Let's merge this PR. Please work with @sudhu2k to review and consolidate `--keep-fp8-transpose-cache`
